src/VbsEnclaveSDK/src/veil_host_lib/hello_winrt_kcm.vtl0.cpp

```diff
@@ -0,0 +1,340 @@
+
+    auto sessionInfo = std::make_shared<uintptr_t>(0);
+    auto enclaveptr = (void*)enclave;   
+
+    try
+    {
+        auto op = KeyCredentialManager::DeleteAsync(key_name);
+        std::wcout << "Deletion worked" << std::endl;
+
+        op.get();
+    }
+    catch (...)
+    {
+        std::wcout << "Deletion failed" << std::endl;
+    }
+    
+    std::wcout << L"Calling RequestCreateAsync" << std::endl;
+    auto credentialResult = KeyCredentialManager::RequestCreateAsync(
+        key_name,
+        static_cast<KeyCredentialCreationOption>(key_credential_creation_option),
+        algorithm,
+        message,
+        cacheConfiguration,
+        static_cast<winrt::Windows::UI::WindowId>(window_id),
+        ChallengeResponseKind::VirtualizationBasedSecurityEnclave,
+        CreateChallengeCallback(sessionInfo, enclaveptr, L"Challenge")
+    ).get();
+
+    std::wcout << L"RequestCreateAsync returned" << std::endl;
+
+    // Check if the operation was successful
+    auto status = credentialResult.Status();
+    if (status != KeyCredentialStatus::Success)
+    {        
+        // Throw after setting result so VTL1 can clean up
+        THROW_HR(static_cast<HRESULT>(status));
+    }
+
+    const auto& credential = credentialResult.Credential();
+
+    // AddRef the COM object to keep it alive when we pass the pointer to VTL1
+    // VTL1 will consume this reference when it's done
+    auto credentialPtr = reinterpret_cast<uintptr_t>(winrt::get_abi(credential));
+    static_cast<IUnknown*>(winrt::get_abi(credential))->AddRef();
+    
+    std::wcout << L"DEBUG: Returning credential pointer: 0x" << std::hex << credentialPtr << std::dec << std::endl;
+
+    credentialAndSessionInfo result;
+    result.credential = credentialPtr;
+    result.sessionInfo = *sessionInfo;
+
+    return result;
+}
+
+veil_abi::Types::credentialAndSessionInfo veil_abi::Untrusted::Implementation::userboundkey_establish_session_for_load(
+    uintptr_t enclave,
+    const std::wstring& key_name,
+    const std::wstring& message,
+    uintptr_t window_id)
+{
+    auto sessionInfo = std::make_shared<uintptr_t>(0);
+    auto enclaveptr = (void*)enclave;
+
+    auto credentialResult = KeyCredentialManager::OpenAsync(
+        key_name.c_str(),
+        ChallengeResponseKind::VirtualizationBasedSecurityEnclave,
+        CreateChallengeCallback(sessionInfo, enclaveptr, L"Load")
+    ).get();
+
+    // Check if the operation was successful
+    auto status = credentialResult.Status();
+    if (status != KeyCredentialStatus::Success)
+    {
+        THROW_HR(static_cast<HRESULT>(status));
+    }
+
+    const auto& credential = credentialResult.Credential();
+
+    // AddRef twice: one for RetrieveAuthorizationContext and one for DeriveSharedSecret
+    auto credentialPtr = reinterpret_cast<uintptr_t>(winrt::get_abi(credential));
+    static_cast<IUnknown*>(winrt::get_abi(credential))->AddRef();
+    static_cast<IUnknown*>(winrt::get_abi(credential))->AddRef();
```

Comment:
memory: I still don't like the double addref.  It's so arcanely specific to this flow, and doesn't currently tolerate any kind of failure w.r.t. cleanup.